include/cetl/span.h

```diff
@@ -1 +1,772 @@
+    using iterator         = pointer;                           //!< Type for iterating through the span.
+    using reverse_iterator = std::reverse_iterator<iterator>;   //!< Reverse iterator type.
+
+    // +----------------------------------------------------------------------+
+    // | Required Properties
+    // +----------------------------------------------------------------------+
+    static_assert(not std::is_pointer<T>::value, "Can not be a pointer type");
+    static_assert(sizeof(element_type) != 0u, "Must have non zero element size");
+
+    // +----------------------------------------------------------------------+
+    ///@{ @name Constructors
+    // +----------------------------------------------------------------------+
+
+    /// Default constructor
+    /// @tparam DeducedExtent   The Extent deduced by the compiler.
+    /// @tparam type            SFNAE enablement of this constructor only where the extent > 0.
+    /// @see std::span::span()
+    template <std::size_t DeducedExtent = Extent, typename std::enable_if<(DeducedExtent == 0), bool>::type = true>
+    constexpr span() noexcept
+        : data_{nullptr}
+    {
+    }
+
+    /// Creates a span starting at an element for a given length.
+    /// The span's view into the data starts at the element pointed to by the first pointer. The size of the
+    /// span is provided by the count where it is undefined behavior to provide a count value that is not the
+    /// same as span::extent.
+    /// @param first The first element in the span.
+    /// @param count The number of elements in the span.
+    /// @see std::span::span()
+    explicit constexpr span(iterator first, size_type count)
+        : data_{first}
+    {
+        (void) count;  // Count isn't actually used in static spans.
+        CETL_DEBUG_ASSERT(count == extent,
+                          "Undefined behavior: Constructing a fixed span where the Extent parameter is different from "
+                          "the count passed into this constructor.");
+    }
+
+    /// Creates a span starting at an element to the element before the given end.
+    /// That is, `span == [first, end)`. It is undefined to provide iterators where `end - first != Extent`
+    /// @tparam EndType End iterator type is deduced to support SFNAE pattern.
+    /// @tparam type Participates in overload resolution only if the EndType cannot be converted to size_type.
+    /// @param first    The first element in the span.
+    /// @param end      The element after the last element in the span.
+    /// @see std::span::span()
+    template <typename EndType,
+              typename std::enable_if<!std::is_convertible<EndType, std::size_t>::value, bool>::type = true>
+    explicit constexpr span(pointer first, EndType end)
+        : data_(first)
+    {
+        (void) end;  // Last isn't actually used in static spans.
+        CETL_DEBUG_ASSERT(std::distance(first, end) == Extent, "Iterator range does not equal Extent.");
```

Comment:
This is about all you can do for static spans but people will complain. Also this will be problematic to detect at runtime on embedded targets (as me how I know).